File: cpp/cpp_100-999/719.cc
// C++
// leetcode 719
// https://leetcode.cn/problems/find-k-th-smallest-pair-distance/

class Solution {
public:
    int smallestDistancePair(vector<int>& nums, int k) {
    	sort(nums.begin(), nums.end());
    	int n = nums.size();
    	int l = 0, r = nums[n-1] - nums[0], mid = 0, cnt = 0;
    	while(l < r) {
    		mid = (l + r) >> 1;
    		cnt = 0;
    		for(int i = 0, j = 0; i < n; i++) {
    			while(j < n && nums[j] - nums[i] <= mid)
    				j++;
    			cnt +=  j - i - 1;
    		}
    		if(cnt >= k) {
    			r = mid;
    		} else {
    			l = mid + 1;
    		}
    	}
    	return r;
    }
};

File: Go/Go_100-999/719.go
import "sort"

// Go
// leetcode 719
// https://leetcode.cn/problems/find-k-th-smallest-pair-distance/

func smallestDistancePair(nums []int, k int) int {
	sort.Ints(nums)
	n := len(nums)
	l, r, mid, cnt := 0, nums[n-1]-nums[0], 0, 0
	for l < r {
		mid = (l + r) >> 1
		cnt = 0
		for i, j := 0, 0; i < n; i++ {
			for j < n && nums[j]-nums[i] <= mid {
				j++
			}
			cnt += j - i - 1
		}
		if cnt >= k {
			r = mid
		} else {
			l = mid + 1
		}
	}
	return r
}